In [1]:
import numpy as np
from scipy.optimize import minimize

In [4]:
# Define the objective function
def objective_function(x):
    return x**3 - 3*x + 1

# Initial guess
x0 = np.array([0])

# Perform the optimization
result = minimize(objective_function, x0)

# Output the result
if result.success:
    optimized_value = result.x[0]
    print(f"Optimized value: {optimized_value}")
    print(f"Function value at optimized value: {objective_function(optimized_value)}")
else:
    print("Optimization failed:", result.message)

Optimized value: 1.0000000040440504
Function value at optimized value: -1.0


In [5]:
# Define the Rosenbrock function
def rosenbrock(x):
    a = 1
    b = 100
    return (a - x[0])**2 + b*(x[1] - x[0]**2)**2

# Initial guess
x0 = np.array([2, 2])

# Minimize using BFGS
res_bfgs = minimize(rosenbrock, x0, method='BFGS')
print("BFGS:", res_bfgs)

# Minimize using Nelder-Mead
res_nelder = minimize(rosenbrock, x0, method='Nelder-Mead')
print("Nelder-Mead:", res_nelder)

BFGS:   message: Optimization terminated successfully.
  success: True
   status: 0
      fun: 1.8932681887175893e-11
        x: [ 1.000e+00  1.000e+00]
      nit: 30
      jac: [ 5.275e-06 -2.506e-06]
 hess_inv: [[ 5.168e-01  1.032e+00]
            [ 1.032e+00  2.066e+00]]
     nfev: 105
     njev: 35
Nelder-Mead:        message: Optimization terminated successfully.
       success: True
        status: 0
           fun: 3.4239552925089766e-10
             x: [ 1.000e+00  1.000e+00]
           nit: 62
          nfev: 119
 final_simplex: (array([[ 1.000e+00,  1.000e+00],
                       [ 1.000e+00,  1.000e+00],
                       [ 1.000e+00,  1.000e+00]]), array([ 3.424e-10,  6.041e-10,  8.954e-10]))


**BFGS Output:**
<br> **message:** "Optimization terminated successfully." This indicates that the optimization process finished without any errors.
<br>**success:** True. The algorithm reached a solution that it considers optimal based on the algorithm's criteria.
<br>status:** 0. This is a code representing the reason the optimizer terminated. In this case, 0 usually means successful completion.
<br>**fun:** The value of the Rosenbrock function at the optimal point found by the algorithm. It is extremely close to 0, which is the minimum value of the function.
<br>**x:** The optimal solution found by the algorithm. It is an array with values [1., 1.], which is the global minimum of the Rosenbrock function at (1,1).
<br>**nit:** Number of iterations. The algorithm took 30 iterations to converge to the optimal solution.
<br>**jac:** The value of the Jacobian (first derivative) of the Rosenbrock function at the optimal point, which is close to zero as expected at a minimum.
<br>**hess_inv:** The inverse of the Hessian matrix at the optimal point. The Hessian matrix provides second-order information (curvature), and its inverse is used in BFGS to update the estimate of the optimal point.
<br>**nfev:** Number of function evaluations. The function was evaluated 105 times.
<br>**njev:** Number of Jacobian evaluations. The Jacobian was evaluated 35 times.
<br>
<br>**Nelder-Mead Output:**
<br>**message:** "Optimization terminated successfully." Like with BFGS, this indicates that the optimization process finished without any errors.
<br>**success:** True. The algorithm found a solution that meets its convergence criteria.
<br>**status:** 0. As with BFGS, this status code indicates success.
<br>**fun:** The final value of the Rosenbrock function, again very close to 0, indicating that the Nelder-Mead method also found the valley.
<br>**x:** The optimal solution found [1., 1.], which is consistent with the BFGS result and confirms the function's global minimum.
<br>**nit:** Number of iterations. The algorithm took 62 iterations, which is more than BFGS. This is typical since Nelder-Mead doesn't use gradient information and can require more iterations.
<br>**nfev:** Number of function evaluations. The function was evaluated 119 times.
<br>**final_simplex:** The last simplex of the algorithm. A simplex in Nelder-Mead is a geometric figure (in this case, a triangle in 2 dimensions) that the method uses to probe the function's landscape. The vertices of the final simplex are very close to the minimum, and the associated function values are given in the array.